x-pack/plugin/ml-feature-index-builder/src/main/java/org/elasticsearch/xpack/ml/featureindexbuilder/action/TransportGetDataFrameJobsStatsAction.java

```diff
@@ -0,0 +1,104 @@
+ */
+
+package org.elasticsearch.xpack.ml.featureindexbuilder.action;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.ActionListenerResponseHandler;
+import org.elasticsearch.action.FailedNodeException;
+import org.elasticsearch.action.TaskOperationFailure;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.tasks.TransportTasksAction;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.metadata.MetaData;
+import org.elasticsearch.cluster.node.DiscoveryNodes;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.discovery.MasterNotDiscoveredException;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.ml.featureindexbuilder.action.GetDataFrameJobsStatsAction.Request;
+import org.elasticsearch.xpack.ml.featureindexbuilder.action.GetDataFrameJobsStatsAction.Response;
+import org.elasticsearch.xpack.ml.featureindexbuilder.job.FeatureIndexBuilderJobTask;
+import org.elasticsearch.xpack.ml.featureindexbuilder.persistence.DataFramePersistentTaskUtils;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class TransportGetDataFrameJobsStatsAction extends
+        TransportTasksAction<FeatureIndexBuilderJobTask,
+        GetDataFrameJobsStatsAction.Request,
+        GetDataFrameJobsStatsAction.Response,
+        GetDataFrameJobsStatsAction.Response> {
+
+    @Inject
+    public TransportGetDataFrameJobsStatsAction(TransportService transportService, ActionFilters actionFilters,
+            ClusterService clusterService) {
+        super(GetDataFrameJobsStatsAction.NAME, clusterService, transportService, actionFilters, Request::new, Response::new,
+                ThreadPool.Names.SAME);
+    }
+
+    @Override
+    protected Response newResponse(Request request, List<Response> tasks, List<TaskOperationFailure> taskOperationFailures,
+            List<FailedNodeException> failedNodeExceptions) {
+        List<DataFrameJobStateAndStats> responses = tasks.stream().map(GetDataFrameJobsStatsAction.Response::getJobsStateAndStats)
+                .flatMap(Collection::stream).collect(Collectors.toList());
+        return new Response(responses, taskOperationFailures, failedNodeExceptions);
+    }
+
+    @Override
+    protected Response readTaskResponse(StreamInput in) throws IOException {
+        return new Response(in);
+    }
+
+    @Override
+    protected void taskOperation(Request request, FeatureIndexBuilderJobTask task, ActionListener<Response> listener) {
+        List<DataFrameJobStateAndStats> jobsStateAndStats = Collections.emptyList();
```

Comment:
This might be simpler if instead of declaring the return value we call the listener once for each branch directly like:

```
listener.onResponse(CollectionsSingletonList(jobStateAndStats)
```

etc. Just a suggestion, I know this is quite subjective :-)